service_container.rst

```diff
@@ -147,64 +167,84 @@ the service container *how* to instantiate it:
+        // thanks to the type-hint, the container will instantiate a
+        // new MessageGenerator and pass it to you!
+        // ...
 
         $message = $messageGenerator->getHappyMessage();
         $this->addFlash('success', $message);
         // ...
     }
 
-When you ask for the ``app.message_generator`` service, the container constructs
-a new ``MessageGenerator`` object and returns it. If you never ask for the
-``app.message_generator`` service during a request, it's *never* constructed, saving
-you memory and increasing the speed of your app. This also means that there's almost
-no performance overhead for defining a lot of services.
+When you ask for the ``MessageGenerator`` service, the container constructs a new
+``MessageGenerator`` object and returns it. But if you never ask for the service,
+it's *never* constructed: saving memory and speed.
 
-As a bonus, the ``app.message_generator`` service is only created *once*: the same
+As a bonus, the ``MessageGenerator`` service is only created *once*: the same
 instance is returned each time you ask for it.
 
+You can also fetch a service directly from the container via its "id", which will
+be its class name in this case::
+
+    use AppBundle\Service\MessageGenerator;
+    
+    // accessing services like this only works if you extend Controller
+    class ProductController extend Controller
+    {
+        public function newAction()
+        {
+            // $messageGenerator = $this->get(MessageGenerator::class);
+
+            $message = $messageGenerator->getHappyMessage();
+            $this->addFlash('success', $message);
+            // ...
+        }
+    }
+
+However, this only works if you set your service to be :ref:`public <container-public>`.
+
 .. caution::
 
-    Service ids are case-insensitive (e.g. ``app.message_generator`` and ``APP.Message_Generator``
-    refer to the same service). But this was deprecated in Symfony 3.3. Starting
-    in 4.0, service ids will be case sensitive.
+    Service ids are case-insensitive (e.g. ``AppBundle\Service\MessageGenerator``
+    and ``appbundle\service\messagegenerator`` refer to the same service). But this
+    was deprecated in Symfony 3.3. Starting in 4.0, service ids will be case sensitive.
 
 Injecting Services/Config into a Service
 ----------------------------------------
 
-What if you want to use the ``logger`` service from within ``MessageGenerator``?
-Your service does *not* have a ``$this->container`` property: that's a special power
-only controllers have.
+What if need to access the ``logger`` service from within ``MessageGenerator``?
+Your service does *not* have access to the container directly, so you can't fetch
+it via ``$this->container->get()``.
 
-Instead, you should create a ``__construct()`` method, add a ``$logger`` argument
-and set it on a ``$logger`` property::
+No problem! Instead, create a ``__construct()`` method with a ``$logger`` argument
+that has the ``LoggerInterface`` type-hint::
```

Comment:
Good suggestion - but I've just done something different: added this note to the controllers article and linked back here :)